src/mongocrypt-ctx-decrypt.c

```diff
@@ -39,68 +59,92 @@ _replace_FLE2IndexedEncryptedValue_with_plaintext (void *ctx,
+      _mongocrypt_key_broker_decrypted_key_by_id (kb, S_KeyId, &S_Key));
+   CHECK_AND_RETURN (mc_FLE2IndexedEncryptedValue_add_S_Key (
+      kb->crypt->crypto, iev, &S_Key, status));
+   const _mongocrypt_buffer_t *K_KeyId =
+      mc_FLE2IndexedEncryptedValue_get_K_KeyId (iev, status);
+   CHECK_AND_RETURN (K_KeyId);
+   CHECK_AND_RETURN_KB_STATUS (
+      _mongocrypt_key_broker_decrypted_key_by_id (kb, K_KeyId, &K_Key));
+   CHECK_AND_RETURN (mc_FLE2IndexedEqualityEncryptedValue_add_K_Key (
+      kb->crypt->crypto, iev, &K_Key, status));
+   const _mongocrypt_buffer_t *clientValue =
+      mc_FLE2IndexedEncryptedValue_get_ClientValue (iev, status);
+   CHECK_AND_RETURN (clientValue);
 
-   if (!_mongocrypt_key_broker_decrypted_key_by_id (kb, S_KeyId, &S_Key)) {
-      _mongocrypt_key_broker_status (kb, status);
-      goto fail;
-   }
+   bson_type_t original_bson_type =
+      mc_FLE2IndexedEncryptedValue_get_original_bson_type (iev, status);
+   CHECK_AND_RETURN (original_bson_type != BSON_TYPE_EOD);
+   CHECK_AND_RETURN_STATUS (
+      _mongocrypt_buffer_to_bson_value (
+         (_mongocrypt_buffer_t *) clientValue, original_bson_type, out),
+      "decrypted clientValue is not valid BSON");
 
-   /* Decrypt InnerEncrypted to get K_KeyId. */
-   if (!mc_FLE2IndexedEncryptedValue_add_S_Key (
-          kb->crypt->crypto, iev, &S_Key, status)) {
-      goto fail;
-   }
+   ret = true;
+fail:
+   _mongocrypt_buffer_cleanup (&K_Key);
+   _mongocrypt_buffer_cleanup (&S_Key);
+   mc_FLE2IndexedEncryptedValue_destroy (iev);
+   return ret;
+}
 
-   const _mongocrypt_buffer_t *K_KeyId =
-      mc_FLE2IndexedEncryptedValue_get_K_KeyId (iev, status);
-   if (!K_KeyId) {
-      goto fail;
-   }
 
-   if (!_mongocrypt_key_broker_decrypted_key_by_id (kb, K_KeyId, &K_Key)) {
-      _mongocrypt_key_broker_status (kb, status);
-      goto fail;
-   }
+static bool
+_replace_FLE2IndexedEncryptedValueV2_with_plaintext (
+   void *ctx,
+   _mongocrypt_buffer_t *in,
+   bson_value_t *out,
+   mongocrypt_status_t *status)
+{
+   bool ret = false;
+   _mongocrypt_key_broker_t *kb = ctx;
+   mc_FLE2IndexedEncryptedValueV2_t *iev =
+      mc_FLE2IndexedEncryptedValueV2_new ();
+   _mongocrypt_buffer_t S_Key = {0};
+   _mongocrypt_buffer_t K_Key = {0};
 
-   /* Decrypt ClientEncryptedValue. */
-   if (!mc_FLE2IndexedEqualityEncryptedValue_add_K_Key (
-          kb->crypt->crypto, iev, &K_Key, status)) {
-      goto fail;
-   }
+   BSON_ASSERT_PARAM (ctx);
+   BSON_ASSERT_PARAM (in);
+   BSON_ASSERT_PARAM (out);
 
+   CHECK_AND_RETURN (mc_FLE2IndexedEncryptedValueV2_parse (iev, in, status));
+   const _mongocrypt_buffer_t *S_KeyId =
+      mc_FLE2IndexedEncryptedValueV2_get_S_KeyId (iev, status);
+   CHECK_AND_RETURN (S_KeyId);
+   CHECK_AND_RETURN_KB_STATUS (
+      _mongocrypt_key_broker_decrypted_key_by_id (kb, S_KeyId, &S_Key));
+   CHECK_AND_RETURN (mc_FLE2IndexedEncryptedValueV2_add_S_Key (
+      kb->crypt->crypto, iev, &S_Key, status));
```

Comment:
I think this will cause an abort if `status` is NULL. We made `status` mandatory in the `add_S/K_Key` functions, but here it's optional.

Reply:
Good catch. Adjusted to create a temporary status as-needed.